common/djangoapps/track/backends/oli.py

```diff
@@ -0,0 +1,121 @@
+"""OLI analytics service event tracker backend."""
+
+from __future__ import absolute_import
+
+import json
+import logging
+
+from requests_oauthlib import OAuth1Session
+
+from django.contrib.auth.models import User
+
+from student.models import anonymous_id_for_user
+from track.backends import BaseBackend
+
+
+LOG = logging.getLogger(__name__)
+
+
+class OLIAnalyticsBackend(BaseBackend):
+    """
+    Transmit events to the OLI analytics service
+    """
+    def __init__(self, **kwargs):
+        super(OLIAnalyticsBackend, self).__init__(**kwargs)
+
+        # the oli analytics service endpoint will be at
+        # "{}{}".format(oli_analytics_service_url, path)
+        self.url = kwargs.get('url', None)
+        self.path = kwargs.get('path', None)
+
+        # shared key for OLI analytics service
+        self.key = kwargs.get('key', None)
+
+        # shared secret for OLI analytics service
+        self.secret = kwargs.get('secret', None)
+
+        self.oauth = OAuth1Session(self.key, client_secret=self.secret)
+
+        # only courses with id in this set will have their data sent
+        self.course_id_set = set(kwargs.get('course_ids', []))
+
+    def send(self, event):
+        """
+        Forward the event to the OLI analytics server
+        Exact API here: https://docs.google.com/document/d/1ZB-qwP0bV7ko_xJdJNX1PYKvTyYd4I8CBltfac4dlfw/edit?pli=1#
+        OAuth 1 with nonce and body signing
+        """
+        if not (self.url and self.secret and self.key):
+            return
```

Comment:
It looks like there are 4 different return values for this method:
- `'OK'` when a 200 response is received from the request
- `'Error'` when a 200 response is _not_ received from the request
- `None` when we use a bare `return` statement (with no return value)
- `None` when we reach the end of the method without executing a `return`
  statement.

Like JavaScript, Python implicitly returns `None` (or `undefined` in JS)
when either:
- a bare `return` statement is used
- a method "falls through" the end without reaching an explicit `return`
  statement

It's confusing when functions mix explicit and implicit return _values_
(`return None` vs. `return`), as well as mixing explicit and implicit
return _statements_ (`return 'OK'` vs. `# function with no return
statement`)

How are these return values used? Looking at other examples of tracking
backends in the platform, I'm not seeing any `return` statements. I'm
not seeing mention in the base class either.

Barring exceptions, if we need to return a value from _a_ code path, we
should explicitly return values from _every_ terminal code path.